File: pufferlib/ocean/flappy_bird_mini/flappy_bird_mini.c
/* FlappyBirdMini: a simple 2-block tall grid environment.
 * Agent can move up or down. Observes an 8x2 grid with moving obstacles.
 * +1 reward per obstacle passed.
 */

#include "flappy_bird_mini.h"

void add_log(FlappyBirdMini* env) {
    env->log.perf += (env->rewards[0] > 0) ? 1 : 0;
    env->log.score += env->rewards[0];
    env->log.episode_length += env->tick;
    env->log.episode_return += env->rewards[0];
    env->log.n++;
}

void c_reset(FlappyBirdMini* env) {
    env->agent_pos = 0; // Start on the floor
    env->wall_pos = rand() % 2;
    env->tick = 0;

    env->observations[0] = (env->wall_pos == 0) ? WALL : EMPTY; // Floor observation
    env->observations[1] = (env->wall_pos == 1) ? WALL : EMPTY; // Ceiling observation
}

void c_step(FlappyBirdMini* env) {
    env->tick += 1;

    int action = env->actions[0];
    env->terminals[0] = 0;
    env->rewards[0] = 0;

    if (action == UP) {
        env->agent_pos = 1;
    } else if (action == DOWN) {
        env->agent_pos = 0;
    }

    // Check for collision with wall
    if (env->agent_pos == env->wall_pos) {
        env->terminals[0] = 1;
        add_log(env);
        c_reset(env);
        return;
    }

    env->observations[0] = (env->wall_pos == 0) ? WALL : EMPTY; // Floor observation
    env->observations[1] = (env->wall_pos == 1) ? WALL : EMPTY; // Ceiling observation

    if (env->tick >= 1000) {
        env->terminals[0] = 1;
        add_log(env);
        c_reset(env);
        return;
    }
}

void c_render(FlappyBirdMini* env) {
    if (!IsWindowReady()) {
        InitWindow(64*2, 64*2, "PufferLib Flappy Bird Mini");
        SetTargetFPS(5);
    }

    // Standard across our envs so exiting is always the same
    if (IsKeyDown(KEY_ESCAPE)) {
        exit(0);
    }

    BeginDrawing();
    ClearBackground((Color){6, 24, 24, 255});

    int px = 64;

    // Draw floor and ceiling
    DrawRectangle(0, px, px*2, px, (Color){100, 100, 100, 255});
    DrawRectangle(0, 0, px*2, px, (Color){100, 100, 100, 255});

    // Draw agent
    Color agent_color = (Color){0, 187, 187, 255};
    if (env->agent_pos == 0) {
        DrawRectangle(px/2, px + px/2, px, px/2, agent_color); // On floor
    } else {
        DrawRectangle(px/2, px/2, px, px/2, agent_color); // On ceiling
    }

    // Draw wall
    Color wall_color = (Color){187, 0, 0, 255};
    if (env->wall_pos == 0) {
        DrawRectangle(px + px/2, px, px/2, px, wall_color); // Wall on floor
    } else {
        DrawRectangle(px + px/2, 0, px/2, px, wall_color); // Wall on ceiling
    }

    EndDrawing();
}

void c_close(FlappyBirdMini* env) {
    if (IsWindowReady()) {
        CloseWindow();
    }
}

int main() {
    FlappyBirdMini env = {};
    env.observations = (unsigned char*)calloc(2, sizeof(unsigned char)); // 2 observations: floor and ceiling
    env.actions = (int*)calloc(1, sizeof(int));
    env.rewards = (float*)calloc(1, sizeof(float));
    env.terminals = (unsigned char*)calloc(1, sizeof(unsigned char));

    c_reset(&env);
    c_render(&env);
    while (!WindowShouldClose()) {
        if (IsKeyDown(KEY_LEFT_SHIFT)) {
            env.actions[0] = NOOP;
            if (IsKeyDown(KEY_UP)    || IsKeyDown(KEY_W)) env.actions[0] = UP;
            if (IsKeyDown(KEY_DOWN)  || IsKeyDown(KEY_S)) env.actions[0] = DOWN;
        } else {
            env.actions[0] = rand() % 3; // 0: NOOP, 1: UP, 2: DOWN
        }
        c_step(&env);
        c_render(&env);
    }
    free(env.observations);
    free(env.actions);
    free(env.rewards);
    free(env.terminals);
    c_close(&env);
}

File: pufferlib/ocean/flappy_bird_mini/flappy_bird_mini.py
"""
FlappyBirdMini: a simple 2-block tall grid environment.
Agent can move up or down. Observes an 8x2 grid with moving obstacles.
+1 reward per obstacle passed.
"""

import gymnasium
import numpy as np

import pufferlib
from pufferlib.ocean.flappy_bird_mini import binding

class FlappyBirdMini(pufferlib.PufferEnv):
    def __init__(self, num_envs=1, render_mode=None, log_interval=128, buf=None, seed=0):
        self.single_observation_space = gymnasium.spaces.Box(low=0, high=1, shape=(16,), dtype=np.uint8) # 8x2 grid flattened = 16 observations
        self.single_action_space = gymnasium.spaces.Discrete(3) # 0: NOOP, 1: UP, 2: DOWN
        self.render_mode = render_mode
        self.num_agents = num_envs
        self.log_interval = log_interval

        super().__init__(buf)
        self.c_envs = binding.vec_init(self.observations, self.actions, self.rewards,
            self.terminals, self.truncations, num_envs, seed)
 
    def reset(self, seed=0):
        binding.vec_reset(self.c_envs, seed)
        self.tick = 0
        return self.observations, []

    def step(self, actions):
        self.tick += 1

        self.actions[:] = actions
        binding.vec_step(self.c_envs)

        info = []
        if self.tick % self.log_interval == 0:
            info.append(binding.vec_log(self.c_envs))

        return (self.observations, self.rewards,
            self.terminals, self.truncations, info)

    def render(self):
        binding.vec_render(self.c_envs, 0)

    def close(self):
        binding.vec_close(self.c_envs)

if __name__ == '__main__':
    N = 4096

    env = FlappyBirdMini(num_envs=N)
    env.reset()
    steps = 0

    CACHE = 1024
    actions = np.random.randint(0, 3, (CACHE, N))

    i = 0
    import time
    start = time.time()
    while time.time() - start < 10:
        env.step(actions[i % CACHE])
        steps += N
        i += 1

    print('FlappyBirdMini SPS:', int(steps / (time.time() - start)))
